src/lib.rs

```diff
@@ -461,3 +463,72 @@ pub trait Format {
 fn default_timestamp() -> u64 {
     0
 }
+
+/// An "adapter" type to feed `Debug` values into defmt macros, which expect `defmt::Format` values
+///
+/// This adapter disables compression! You should prefer `defmt::Format` over `Debug` whenever
+/// possible
+///
+/// This adapter works by formatting the `Debug` value into a stack-allocated buffer. You need to
+/// specify how large that buffer is. If you pick a size that's too small an *incomplete* string
+/// will be transmitted.
+///
+/// The size of the stack-allocated array is specified using one of the type-level integers in the
+/// `consts` module. Example:
+///
+/// ```
+/// use defmt::{consts, Debug2Format};
+///
+/// #[derive(Debug)]
+/// struct S { x: i8, y: i16 }
+///
+/// defmt::info!("{:?}", Debug2Format::<consts::U32>(&S { x: 1, y: 2 }));
+/// //                                  ^^^^^^^^^^^ size = 32 bytes
+/// // -> 0.000000 INFO  S { x: 1, y: 2 }
+/// ```
+pub struct Debug2Format<'a, N>
+where
+    N: ArrayLength<u8>,
+{
+    // we don't store a heapless String inline because rustc is not great at optimizing moves of
+    // large arrays and that can result in double stack usage
+    _capacity: PhantomData<N>,
+
+    // we use a trait object here to avoid adding a second type paratemer to `Debug2Format`. if
+    // `Debug2Format` has 2 type parameter then users need to add an underscore type parameter to
+    // the turbofish: `Debug2Format::<U64, _>(value)`
+    value: &'a dyn fmt::Debug,
+}
+
+impl<N> Format for Debug2Format<'_, N>
+where
+    N: ArrayLength<u8>,
+{
+    fn format(&self, fmt: &mut Formatter) {
+        use core::fmt::Write as _;
+
+        // `Debug` format into a stack buffer; if the formatted string doesn't fit discard the
+        // excess
+        let mut buf = String::<N>::new();
+        write!(buf, "{:?}", self.value).ok();
```

Comment:
Could we maybe use an ellipsis here? Maybe put a `…` at the end of the buffer if the full text doesn't fit?

Reply:
I can't think of a nice way to do that in a way that ensures the ellipsis will be included if the output is truncated.
Reason being the `write!` may fill `buf` with truncated output. In that scenario we cannot push the ellipsis at the end (because the buffer is full).

> I can't think of a nice way

But I can think of not nice ways to do this. One is passing `String<N - 2>` to `write!` by transmuting a mutable reference (yeah, let's not do that). The other is to use 2 heapless String of sizes N and N + 2: `write!` into one; if that's truncated then memcpy into the larger one + add ellipsis. The last way is to write our own heapless string that has internal capacity N+2 but looks like it has capacity N to the end user; that string would have some method to append an ellipsis.

P.S. The heapless::String's fmt::Write implementation is a bit odd in that it won't write an argument at all if it doesn't fit. That is `write!(s, "x = {}", u32::max_value())` produces "x = " instead of "x = 4294" if the string capacity is 8.

Reply:
Ah, I see, yeah let's just land this as-is then.